lib/index.js

```diff
@@ -479,6 +488,83 @@ Socket.prototype.send = function(msg, flags) {
   return this;
 };
 
+/**
+ * Read one message from socket.
+ *
+ * @return {Array} multi-part message or null if socket doesn't have any
+ * @api public
+ */
+Socket.prototype.read = function() {
+  var message = [], flags;
+
+  if (this._zmq.state !== zmq.STATE_READY) {
+    return null;
+  }
+
+  flags = this._zmq.getsockopt(zmq.ZMQ_EVENTS);
+
+  if (flags & zmq.ZMQ_POLLIN) {
+    do {
+      message.push(this._zmq.recv());
+    } while (this._zmq.getsockopt(zmq.ZMQ_RCVMORE));
+
+    return message;
+  }
+
+  return null;
+}
+
+/**
+ * Send the given `msg`.
+ *
+ * @param {String|Buffer|Array} msg
+ * @param {Function} callback function (optional)
+ * @return {Socket} for chaining
+ * @api public
+ */
+Socket.prototype.write = function (parts, callback) {
+  var self = this,
+    i = 0,
+    len = parts.length,
+    flags = this._zmq.getsockopt(zmq.ZMQ_EVENTS),
+    part;
+
+  if (!Array.isArray(parts)) {
+    parts = [parts];
+  }
+
+  // socket is not ready to accept messages
+  if (!(flags & zmq.ZMQ_POLLOUT)) {
+    server.once('ready', function () {
+      self.write(parts, callback);
+    });
```

Comment:
Sure, we on our end broke the _outgoing regularly :). This is not to called lightly. In writable stream case buffering is happening in stream.Writable class and it is calling **write** in sequence mode, we would need a pref test to see direct impact. 

I'm using 0mq with 100k messages per sec in a distributed system, we have had mem leaks on push and pull side, so I would say we managed to break both ends. More specific **_outgoing** and eventEmitter are the sources of leak. 

We already have this PR in production and it is solving the problem on pull side by piling up data in 0mq and not in heap. Potential downside that I saw, but I can not guaranty, is that it seems to me that system load went up due to constant switching between node and 0mq...

We did not use Writable side in production yet.